book/src/dev/rfcs/0005-treestate.md

```diff
@@ -0,0 +1,131 @@
+a note commitment tree state given the assumed security properties of the Merkle
+tree’s hash function.  Since the nullifier set is always updated together with
+the note commitment tree, this also identifies a particular state of the
+associated nullier set.
+
+**spend descriptions**: A shielded Sapling transfer that spends a note. Includes
+an anchor of some previous block's note commitment tree.
+
+**output descriptions**: A shielded Sapling transfer that creates a
+note. Includes the u-coordinate of the note commitment itself.
+
+**joinsplit**: A shielded transfer that can spend Sprout notes and transparent
+value, and create new Sprout notes and transparent value, in one Groth16 proof
+statement.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As blocks are validated, the note commitments revealed by all the transcations
+within that block are used to construct note commitment trees, with the note
+commitments aligned in their note positions in the bottom layer of the Sprout or
+Sapling note tree from the left-most leaf to the right-most in transaction order
+in the block. So the Sprout note commitments revealed by the first
+JoinSplit<Groth16> in a block would take note position 0 in the Sprout note
+commitment tree, for example. Once all the transactions in a block are parsed
+and the notes for each tree collected in their appropriate positions, the root
+of each tree is computed. While the trees are being built, the respective block
+nullifier sets are updated in memory as note nullifiers are revealed. If the
+rest of the block is validated according to consensus rules, that root is
+committed to its own datastructure via our state service (Sprout anchors,
+Sapling anchors). Sapling block validation includes comparing the specified
+FinalSaplingRoot in its block header to the root of the Sapling note commitment
+tree that we have just computed to make sure they match.
+
+As the transactions within a block are parsed, Sapling shielded transactions
+including Spend descriptions and Output descriptions describe the spending and
+creation of Zcash Sapling notes, and JoinSplit on Groth16 descriptions to
+transfer/spend/create Sprout notes and transparent value. JoinSplit and Spend
+descriptions includes an anchor, which references a previous note commitment
+tree root: for Spend's, this is a previous block's anchor as defined in their
+block header, for JoinSplit's, it may be a previous block's anchor or the root
+produced by a previous JoinSplit transfer in its transaction. For Spend's, this
+is convenient because we can query out state service for previously finalized
+Sapling block anchors, and validate Spend description independently. For
+JoinSplit's, if it's not a previously finalized block anchor, it must be the
+treestate anchor of previous JoinSplit in this transaction, and we have to wait
+for that one to be parsed and its root computed to check that ours is valid.
```

Comment:
@hdevalence should we save intra-transaction joinsplit anchors down to sled? I think no. How should we shuffle this dependent data while we're in the middle of validating a block?

Reply:
I agree no, you should not need to persist that data.